File: LeetCode/iter1/python/HouseRobberII.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        lstSize = len(nums)
        if(lstSize == 0):
            return 0
        elif(lstSize == 1):
            return nums[0]
        else:
            #包含第一个，去除最后一个的
            lwithFirst = [0,nums[0]]
            for i in range(2,lstSize):
                lwithFirst.append(max(lwithFirst[i-2]+nums[i-1],lwithFirst[i-1]))
            
            #包含最后一个，不包含第一个
            lwithoutFirst = [0,nums[1]]
            for i in range(2,lstSize):
                lwithoutFirst.append(max(lwithoutFirst[i-2]+nums[i],lwithoutFirst[i-1]))
                
            # print(lwithFirst)
            # print(lwithoutFirst)
            return max(lwithFirst[lstSize-1],lwithoutFirst[lstSize-1])
            
            
#测试程序
s = Solution()
print(s.rob([1,2,3,4,5]))

File: LeetCode/iter2/C++/HouseRobberII.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int rob(vector<int>& nums) {
	if (nums.size() == 0)
		return 0;

	if (nums.size() == 1)
		return nums[0];

	if (nums.size() == 2)
		return max(nums[0],nums[1]);

	int *dp = new int[nums.size()];
	dp[0] = 0;
	dp[1] = nums[1];

	//不包含第一个，包含最后一个
	for (int i = 2; i < nums.size(); i++)
		dp[i] = max(dp[i-2]+nums[i],dp[i-1]);

	int candidate = dp[nums.size() - 1];

	//包含第一个，不包含最后一个
	dp[0] = nums[0];
	dp[1] = max(nums[0],nums[1]);
	for (int i = 2; i < nums.size() - 1; i++)
		dp[i] = max(dp[i-2]+nums[i],dp[i-1]);

	return max(candidate,dp[nums.size()-2]);
}


int main()
{
	vector<int> v;
	v.push_back(2);
	v.push_back(4);
	v.push_back(3);
	v.push_back(4);
	v.push_back(5);

	cout << rob(v)<<endl;

	vector<int> v1;
	v1.push_back(2);
	v1.push_back(2);
	v1.push_back(4);
	v1.push_back(3);
	v1.push_back(2);
	v1.push_back(5);

	cout << rob(v1)<<endl;

	vector<int> v2;
	v2.push_back(2);
	v2.push_back(1);
	v2.push_back(1);
	v2.push_back(1);

	cout << rob(v2) << endl;

	vector<int> v3;
	v3.push_back(1);
	v3.push_back(1);
	v3.push_back(2);
	v3.push_back(1);

	cout << rob(v3) << endl;

	return 0;
}